### PR TITLE
SearchControl: Convert to TypeScript and add flag to remove bottom margin

### PR DESCRIPTION
**Pull request by mirka**
Part of #38730 #35744 

## What?

- Converts SearchControl to TypeScript.
- Adds the `__nextHasNoMarginBottom` prop to remove the bottom margin.

## Why?

Better docs, better reusability.

## Testing Instructions

`npm run storybook:dev` and check the SearchControl stories in Docs view.

### CODE REVIEW

File: packages/components/src/search-control/index.tsx [hunk 1]
```diff
@@ -72,6 +78,7 @@ function SearchControl(
 		>
 			<div className="components-search-control__input-wrapper">
 				<input
+					{ ...restProps }
```
Comment (mirka) [inline packages/components/src/search-control/index.tsx:81]: ⚠️ Runtime change: Pass down rest props

File: packages/components/src/search-control/stories/index.tsx [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	onClose,
+	...props
+} ) => {
+	const [ value, setValue ] = useState< string >();
+
+	return (
+		<SearchControl
+			{ ...props }
+			value={ value }
+			onClose={ onClose }
```
Comment (ciampo) [inline packages/components/src/search-control/stories/index.tsx]: Can the `onClose` prop be passed together with `...props` ?

File: packages/components/src/search-control/README.md [hunk 3]
```diff
@@ -48,12 +48,13 @@ If this property is added, a specific placeholder will be used for the input.
 
 -   Type: `String`
 -   Required: No
+-   Default: `__( 'Search' )`
```
Comment (ciampo) [inline packages/components/src/search-control/README.md:51]:
```suggestion
-   Default: `__( 'Search' )`

```

**Review by ciampo — COMMENTED** [review]
Testing the component in Storybook, pressing the "X" button doesn't seem to be clearing the input's value

File: packages/components/src/search-control/stories/index.tsx (hunk 2 above)
Comment (mirka) [reply packages/components/src/search-control/stories/index.tsx]: 🙈

**Comment by mirka** [comment]
> Testing the component in Storybook, pressing the "X" button doesn't seem to be clearing the input's value

Ah thanks, I messed that up.

In 525dd66 I fixed it so the `onClose` action spy is only passed in the `onClose` story. I also improved the story/prop description a bit to clarify its usage.

**Review by ciampo — APPROVED** [review]
🚀 

> In https://github.com/WordPress/gutenberg/commit/525dd66264ba04dfaf7d4191d2ccff72f1fdbdb0 I fixed it so the onClose action spy is only passed in the onClose story. I also improved the story/prop description a bit to clarify its usage.

That is an unexpected behavior that the component is having 🤔  but anyway, not the scope of this PR.

**Pull request merged**